### ISSUE TITLE
World Hangs forever when generating Underground Ruins in oceans when Zombie Horse Spawn is installed [1.15.2]

### ISSUE THREAD

**Issue by NJL5390** [svenhjol/Strange#40]
When testing a custom modpack on a server I've had set up for a while, I encountered a bug where the world would keep freezing then kicking me out of the server. I wouldn't be able to re-join after that and I would have to use task manager to close the server and delete the world files in order to use the server again. 

After extensive testing in both singleplayer and multiplayer I found from the server's debug log that the issue occurred whenever the world tried to generate a type of ruin called "ocean_chambers". Removing Strange from my mod list seemed to fix the issue as well. The line in the log that displayed this was:
[26Apr2020 22:34:32.934] [Server-Worker-91/DEBUG] [strange/INTERNAL]: [UndergroundStructure] create ruin ocean_chambers at BlockPos{x=-3120, y=20, z=-2464}

The full debug log is below:
[debug.log](https://github.com/svenhjol/Strange/files/4536552/debug.log)

**Comment by NJL5390** [comment]
Here is the full mod list for the server in case you need it:
Aquaculture - 2.0.13
Atmospheric - 1.2.1
AutoRegLib - 1.5-40
Autumnity - 1.2.0
Bambooblocks - 1.3
Berry Good - 2.0.0
Biomesoplenty - 10.0.0.355
Bloomful - 1.3.2
BonsaiTrees - 2.1.2.6
Bounty of the land - 0.1.2
Breed killer rabbit - 1.1
Buzzier Bees - 1.4
Caelus - FORGE - 2.0-beta2
Campfire torches - 1.0.2
Charm - 1.5.7
Clumps - 5.0.2
Colytra - 3.0
Comforts - 2.0.0.1
Cooking for Blockheads - 8.0.1
Corail woodcutter - 1.1.2
Cosmetic Armor reworked - v2
Fast Furnace - 3.0.0
Fast leaf decay - v22
Fast Workbench - 3.0.0
Guard Villagers - 1.0.6
Hwyla - 1.10.8-B72
Incubation - 1.0.3
Kiwi - 2.7.0
Libnonymus - 1.1.1.5
Login Protection - 1.2
Macaw's bridges - 1.0.4
Mixin Bootstrap - 1.0.3
Mooshroom Tweaks - 1.0
Mystical lib - 2.0.0
Naturally Spawning Vex - 1.0.0.4
Naturally Charged creepers - 1.1
Pam's HC2 crops - 1.0.3
Pam's HC2 food core - 1.0.2
Pam's HC2 trees - 1.0.3
Performant - 2.09m
Phosphor - 0.5.2+build4
Placebo - 3.0.2
Polarizing Biomes - v1.0
Quark Oddities - 1.15.2
Quark - r2.1-234
Random Patches - 1.21.1.2
Repurposed Structures - 1.3.0
Respawnable pets - 1.1
Save my Stronghold - 1.0.0
Simple Farming - 1.2.7
Skeleton Horse Spawn - 1.1.2
Snow Real Magic - 1.8.0
Snow under trees - 1.0
Spice of life: Carrot edition - 1.9.4
Strange - 1.0.0-beta3
Swamp Expansion - 1.5.1
Swing through grass - 1.4.0
The Endergetic Expansion - 1.2.8
Corail Tombstone - 4.3.5
Traverse - 3.0.0
Upgrade Aquatic - 1.5.1
Valhelsia Structures - 15.0.2a
Wither Skeleton Tweaks - 4.0.0
Zombie Horse spawn - 2.1.2

Forge Version for the server is: 31.1.12 (Though the issue does still occur with more recent versions)

**Comment by svenhjol** [comment]
Thanks for the report. Unfortunately that log doesn't help me debug the problem. Please can you try and repeat with just Strange and Charm in a copy of your world, and let me know what happened when it crashed.

**Comment by NJL5390** [comment]
When I loaded the copy of the world up and joined the server via my client, I was able to move around, within the chunk I was in for about 10 seconds (no other chunks loaded in) before the server crashed and kicked me back to the multiplayer menu. All entities near me were frozen in place while this was happening.

**Comment by svenhjol** [comment]
Did you make any changes to your strange config file? If so can you paste it here

**Comment by NJL5390** [comment]
No, I have not made any changes to the config. 

I have tried to reproduce this issue with only Charm and Strange and I was unable to. I visited many different types of ruins (including some ocean ones) and some vaults and stone circles and nothing seemed to cause an issue.

**Comment by svenhjol** [comment]
You have something in your log that is causing an enormous amount of lag.
```
[26Apr2020 22:30:36.904] [Server thread/INFO] [com.performant.coremod.Performant/]: Pausing entity AI due to lag, last 20 ticks took:299ms. Stacktrace below, let the mod author know(not performant, the entities mod author)
java.lang.Exception: Entity AI Task lag
```

This error's stacktrace includes this line:
`tallestegg.guardvillagers.entities.GuardEntity.func_70636_d(GuardEntity.java:128) ~[?:1.15.2-1.0.6]`

I'm not familiar with what this mod is, but I'd definitely try adding that one to Strange and Charm to see if it's lagging your world.

**Comment by NJL5390** [comment]
Guard Villagers is the mod you are referring to: https://www.curseforge.com/minecraft/mc-mods/guard-villagers

The issue with ruins and world freezing/hanging never seems to occur around a village (Which is what this mod changes), it always seems to occur when I'm near an ocean (Every other type of ruin I generate works as it should). Which is weird when considering this particular error. I only notice this AI lag when I see it in the log.

I was going to report this AI lag issue to that mod's author but I'm a bit short on time at the moment.

However I will test this mod alongside Charm and Strange and see if any issues arise. Lag or otherwise.

**Comment by NJL5390** [comment]
From what testing I have been able to do, I can say that "Guard Villagers" does not cause the issue with Underground ocean ruins.

In terms of lag it is definitely causing some, but is only really noticeable when teleporting using Runestones or loading and generating new chunks or when the guards are attacking mobs. Most of the time there doesn't really seem to be anything wrong in terms of performance.

**Comment by NJL5390** [comment]
On further testing I have narrowed the list above down to 6 mods which can potentially cause the issues mentioned above:
Naturally Spawning Vex - 1.0.04
Naturally Charged Creepers - 1.1
Skeleton Horse Spawn - 1.1.2
Zombie Horse Spawn - 2.1.2
Mooshroom Tweaks - 1.0
Breed Killer Rabbit - 1.1

I have tested generating these ocean ruins with only Charm and Strange as well as these mods and found I was able to replicate the issue with Underground Ocean ruins. When I remove these 6, the issue disappears and everything works normally.

Another thing that I have noticed when testing with just these 6 mods (+Charm and Strange) is that these mods appear to cause considerable entity lag as well, which also disappears when I remove these mods. (That includes the thing you noticed in my log causing enormous lag, that disappears as well).

**Comment by svenhjol** [comment]
Thanks for researching this, it's a big help. I can tell you the underground ruins won't spawn entities until the player enters the ruin. The entities that spawn in the underwater ruins are drowned and squid only. None of the above mods look like they might have an issue with ruins based solely on their name.

**Comment by NJL5390** [comment]
So, I've done a bit more testing and I have found that the one mod that is causing this issue is Zombie Horse Spawn (2.1.2).

When I've tested without this mod, everything works fine (even with all other mods included). When I test this by itself (With Charm and Strange) the issue occurs. 

What I can tell you about this mod is that the only thing it does is give all zombies (and their variants) a 1/20 chance (5%) of spawning riding a zombie horse (I changed this value in the config to 1/10). What I have noticed with more recent tests is that the world only starts to freeze when I enter the ocean ruins themselves (which is, as you said when the ruins start spawning mobs). What I theorize could be happening is that since the drowned is a variant of the zombie, the mod will try and give it a zombie horse every so often which then leads to this world freezing issue whenever the spawners in the underground ruins try to spawn a drowned.

That's just my theory though, I don't know what could be going on within this mod to cause this issue other than that.

**Comment by svenhjol** [comment]
This gives me a lot more to work with. Thank you!

**Comment by NJL5390** [comment]
I got time to do some more testing on this issue lately, So I tested messing around with the config for Zombie Horse Spawn and nothing was making any sort of difference, I then decided I would delete and regenerate the config file for the mod with the newest version of forge and see if that made a difference.

Well, It did make a difference in that it made this issue far rarer and more inconsistent, it was still only occurring near oceans but I was actually able to generate several underground ocean ruins without problems.

I then decided I would test with only Zombie Horse Spawn and see if this issue ever occurred near vanilla ocean structures, and I found that it did, but it happened very inconsistently and or after I spent a certain amount of time in the world. 

Therefore, I'm beginning to get the impression that this issue isn't actually caused by Strange and could just be caused by Zombie Horse Spawn itself. I'm unsure though, since this issue does not happen regularly and can be quite erratic, but since the world freezing issue also seems to occur near vanilla structures (Such as Ocean ruins and monuments) without Strange or Charm installed, I would guess that this issue isn't related to Strange.

**Comment by svenhjol** [comment]
Scheduled for testing in the next beta.

**Comment by svenhjol** [comment]
Updated thread on their issue tracker. Moving this ticket out of beta 5 for future release.

**Comment by NJL5390** [comment]
Fixed as of Zombie Horse Spawn v2.3 and Collective v1.8
